### PR TITLE
[WIP] add initial support for StorageVersion API for CRDs

### PR DESCRIPTION
**Pull request by andrewsykim**
Signed-off-by: Andrew Sy Kim <andrewsy@google.com>
Co-authored-by: Haowei Cai (Roy) <haoweic@google.com>

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR is based off of https://github.com/kubernetes/kubernetes/pull/96403, but rebased off of master. Implementation still needs a lot of work, opening the PR now to get more feedback and run CI against the changes.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
add initial support for StorageVersion API for CRDs
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by apelisse** [comment]
The KEP is multi-year old, are we still planning on doing that? I'd rather see an api-machinery TL green light before I spend time reviewing this?

**Comment by andrewsykim** [comment]
> The KEP is multi-year old, are we still planning on doing that? I'd rather see an api-machinery TL green light before I spend time reviewing this?

@apelisse please don't review the PR now, it's still WIP. But you can see the latest KEP update for v1.26 if you're interested https://github.com/kubernetes/enhancements/pull/3593

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

**Comment by dims** [comment]
This PR has the label work-in-progress, please revisit to see if you still need this, please close if not

File: staging/src/k8s.io/apiserver/pkg/storageversion/updater.go [hunk 1]
```diff
@@ -122,13 +125,23 @@ func setStatusCondition(conditions *[]v1alpha1.StorageVersionCondition, newCondi
...
 
-// updateStorageVersionFor updates the storage version object for the resource.
-func updateStorageVersionFor(c Client, apiserverID string, gr schema.GroupResource, encodingVersion string, decodableVersions []string) error {
+// UpdateStorageVersionFor updates the storage version object for the resource.
+// postProcessFunc is invoked on the storage version object after the local
+// calulation is finished, and before the client sends the create/update request.
+// It allows the caller to customizes the storage version object, e.g. setting
+// an owner reference.
+func UpdateStorageVersionFor(ctx context.Context, c Client, apiserverID string, gr schema.GroupResource, encodingVersion string, decodableVersions []string, postProcessFunc processStorageVersionFunc, alwaysRetry bool) error {
 	retries := 3
```
Comment (Richabanker) [inline staging/src/k8s.io/apiserver/pkg/storageversion/updater.go:134]: Do we want to make this configurable @andrewsykim ?

**Pull request closed without merging**